File: Android/src/main/java/com/cisco/or/sdk/services/ServerAuthCodeService.kt
package com.cisco.or.sdk.services

import android.content.Context
import com.cisco.or.sdk.types.ServiceHandler
import com.cisco.or.sdk.utils.Urls
import com.cisco.or.sdk.utils.Utils
import org.json.JSONObject
import kotlin.collections.ArrayList

/**
 * Only OAuth Token
 */
internal object ServerAuthCodeService : Service() {
    private val TAG: String = ServerAuthCodeService::class.java.name

    private var oAuthToken : String = ""
    private var serviceName : String = ""

    /**
     * Not implemented.
     * Use start(context: Context, oAuthToken: String, serviceName: SigningServiceName, serviceHandler: ServiceHandler).
     */
    override fun start(context: Context, serviceHandler: ServiceHandler) {
        if(Utils.internalCall(Service::class.qualifiedName.toString(), "start")){
            super.start(context,serviceHandler)
        }else{
            throw NotImplementedError()
        }
    }

    fun start(context: Context, oAuthToken: String, serviceName: String, serviceHandler: ServiceHandler) {
        this.oAuthToken = oAuthToken
        this.serviceName = serviceName
        super.start(context, serviceHandler)
    }

    override fun onStart(context: Context, headerParams: ArrayList<HeaderParam>, serviceHandler: ServiceHandler) {
        val body = JSONObject()
        body.put("authType", this.serviceName)
        body.put("id", this.oAuthToken)

        val url = Urls.oauthTokenUrl()
        val requestData = RequestData(url, headerParams, HTTP_METHOD.POST, body.toString())
        call(context, requestData, RESPONSE_FORMATTER.TEXT, serviceHandler)
    }
}

File: IOS/Sources/OpenRoaming/OpenRoaming/Services/ServerAuthCodeService.swift
//
//  ServerAuthCodeService.swift
//  OpenRoaming
//
//  Created by Luis Vasquez on 16/06/20.
//  Copyright © 2020 Cisco. All rights reserved.
//

import Foundation

class ServerAuthCodeService: Service {
    var serverAuthCode: String!
    var serviceName: ServiceName!
    
    override func start(serviceHandler: @escaping ServiceHandler) {
        fatalError("Not implemented")
    }
    
    func start(serverAuthCode: String, serviceName: ServiceName, serviceHandler: @escaping ServiceHandler) {
        self.serverAuthCode = serverAuthCode
        self.serviceName = serviceName
        super.start(serviceHandler: serviceHandler)
    }
    
    override func onStart(headerParams: Array<HeaderParam>, serviceHandler: @escaping ServiceHandler) {
        let url = Urls.oauthTokenUrl()
        
        let jsonValues:NSMutableDictionary = NSMutableDictionary()
        jsonValues.setValue(self.serviceName.rawValue, forKey: "authType")
        jsonValues.setValue(self.serverAuthCode, forKey: "id")
        
        let jsonData = try! JSONSerialization.data(withJSONObject: jsonValues)
        let body = NSString(data: jsonData, encoding: String.Encoding.utf8.rawValue)! as String
 
        let requestData = RequestData(url: url, headerParams: headerParams, method: HTTPMethod.POST, body: body)
        call(requestData: requestData, serviceHandler: serviceHandler)
    }
}
